PASS Developer Guidelines (#13)
eclipse-pass/pass-documentation: community/developer-guidelines.md

* We primarily use Slack to communicate about PASS development. To be invited to our Slack workspace please send us a message on our [Google Group](https://groups.google.com/g/pass-general).
* Contributing to the project begins as a [contributor](https://www.eclipse.org/projects/handbook/#contributing-contributors) and may lead to being a [committer](https://www.eclipse.org/projects/handbook/#roles-cm). Whether you are a `contributor` or `committer` you will need to sign up for an [Eclipse account](https://accounts.eclipse.org/user/login).
    * A `contributor` can add to and improve PASS by creating issues and submitting pull requests. You’ll find more information about both of these tasks below.
    * A `committer` is an individual who once was a `contributor`, but made significant contributions and was elected by the core team to become a `committer`. They can work directly in the repositories, create and close issues, and merge pull requests.

> * Read through the PASS Community Guidelines.

Author: I agree, it does sound better with your suggestion. Since the community guidelines is not developed at the moment (ticket in the backlog) I will remove this bullet and make note to add it back in once the community guideline is completed.

## Change Request/Bug Report

> The process for initiating a change request or submitting a bug report is through submitting a GitHub issue.

Author: Done.

* When creating an issue to report a bug or suggest a new feature, use the [eclipse-pass/main repository](https://github.com/eclipse-pass/main/issues).
* If available for your particular issue use one of the available [issue templates in the main repository](https://github.com/eclipse-pass/main/issues/new/choose).
* If a template doesn’t exist, use the `Standard Issue` template.
* Add a label if possible, if the label doesn’t exist, or you’re unsure of which one to use, send a message to the  team on the Slack `#pass-dev` channel.
* If possible, suggest a priority. If unsure of the priority, leave it blank and the team will decide what is the priority for the issue.

## Testing

> * It is suggested to run tests locally (`mvn verify` or `mvn clean install`) and [run pass docker](../welcome-guide/setup-run-pass.md) to test.

Author: Done.

> * For a general idea of how tests should be structured, see Martin Fowler’s [The Practical Testing Pyramid](https://martinfowler.com/articles/practical-test-pyramid.html).

Author: Done.

### Back-end

> 

Reviewer: I'd suggest adding a very brief introduction to the types of tests reference here (e.g. unit tests are ...) so everyone knows what we mean when we say "Acceptance Tests"

Author: Done. I used the resource by martin fowler for these. There are nuances to these tests as mentioned by fowler, but I think what I wrote aligns with our team's definition.

Reviewer: The Unit Test and Acceptance Tests sections look good (though I'd suggest a tweak to clarify the last sentence of the Unit test section: `Unit tests alone do not guarantee the application as a whole will work as intended.`) The Integration Test section doesn't quite fit with my understanding of how integration tests are used in PASS. My take is closer to the definition here: https://www.geeksforgeeks.org/software-engineering-integration-testing/.

Author: The definition for integration tests was taken from Martin Fowler's `The Practical Test Pyramid` resource that we cite in the developer guide: " you're planning to submit code which includes new tests, Martin Fowler’s [The Practical Testing Pyramid](https://martinfowler.com/articles/practical-test-pyramid.html) is a great resource for understanding how to structure tests." This was a resource suggested by the team. Should we revisit this or do you think it would be best to remove it altogether?

Author: I think our integration tests are a bit of a mix between the two definitions. There are certainly cases where we have integration tests where we use testcontainers e.g. [simulate an S3 bucket and test the integration with loading a policy json file ](https://github.com/eclipse-pass/pass-support/blob/626668c7ed295f764d788dd99bd0994d3a038f24/pass-grant-loader/src/test/java/org/eclipse/pass/support/grant/PolicyPropertiesS3IT.java)

I can revise this as well to be a mix of the two definitions?

Author: I do agree that our ITs are not the exact definition given by Fowler. He does get a bit more nuanced later on, by describing contract and providers tests. I think we have a more broad definition of an IT than Fowler does.

Author: As discussed, I updated with a merged definition and provided the resource you mentioned. This is now done.

* **Unit Tests**

>     * If introducing new functionality, this should be tested, including at minimum two pathways: success and failure states.

Author: Done.

        * A few examples of this are [here](https://github.com/eclipse-pass/pass-support/blob/79ad19ed4d2592c342e7cdfdf652a8f7aef3eaa2/pass-deposit-services/deposit-core/src/test/java/org/eclipse/pass/deposit/service/DepositProcessorIT.java#L54) and [here](https://github.com/eclipse-pass/pass-core/blob/e9e853ac7eea05f595fdcd5342ddea99c0798e38/pass-core-main/src/test/java/org/eclipse/pass/object/ElidePassClientTest.java#L84).
    * If performing a bug fix, include a test to ensure that the bug was fixed.
    * In general, unit tests should run quickly.
* **Integration Tests**

>     * Set up a test environment that mimics the production environment

Author: Done.

>     * Avoid making network requests to 3rd parties.

Author: Done.

         * If needed, use test containers, wiremock, or mockbean.
    * Integration tests should be as fast as possible.
* **Acceptance Tests**

>     * Should be based on user requirements and work correctly from a user’s perspective.

Reviewer: ```suggestion
    * Acceptance tests are used in PASS to verify correct functionality based on user requirements, so these should work correctly from a user’s perspective.
```

Author: Done.

    * Updated whenever there are changes to user requirements, significant changes are made to the application, or when they break.
    * Automated so they can be run frequently and consistently.

### UI

* When testing the UI it is helpful to run [Ember locally for faster iteration](https://github.com/eclipse-pass/main/blob/main/docs/dev/running-pass-ui-on-your-host-machine.md).

> * Unit test when you can, when functions don't interact with rendering, but otherwise utilize component integration or ember application/acceptance tests where rendering is involved - this is what ember is best at.

Reviewer: ```suggestion
* Include unit test when you can, such as when functions don't interact with rendering. Otherwise, utilize component integration or ember application/acceptance tests where rendering is involved - this is what ember is best at.
```

Author: Done.

* Application level testing is done with mocked data using Mirage. This needs to be updated diligently, so it doesn't fall out of sync with the real back end. If you are updating the API in a way that changes the contract with pass-ui, please at least create an issue for updating the UI mocking to accommodate these changes.
* At least one test should be added for bug fixes to prevent regression.
* The [pass-acceptance-testing](https://github.com/eclipse-pass/pass-acceptance-testing) acceptance tests are run frequently and can aid in making sure the UI application testing mocked responses are appropriate.
* Ember provides a set of very helpful libraries to assist in testing UI components. It is best practice to use these helpers rather than inventing your own where possible:
    * [Ember test-helpers](https://github.com/emberjs/ember-test-helpers/blob/master/API.md) 
    * [QUnit-dom](https://github.com/mainmatter/qunit-dom/blob/master/API.md) 

## Commits

* As a team, we do not enforce rigid rules about commit messages. However, we strive to write good commit messages by following these [guidelines by Chris Beams](https://cbea.ms/git-commit/).

## Documentation

> * Read through the [PASS documentation style guide](https://docs.google.com/document/d/11aCooQCNhEq34yG9mGuFynY4xMDRJtPRuOOou9LaCgU/edit?usp=sharing).

Reviewer: ```suggestion
* We encourage you to read through the [PASS documentation style guide](https://docs.google.com/document/d/11aCooQCNhEq34yG9mGuFynY4xMDRJtPRuOOou9LaCgU/edit?usp=sharing) prior to submitting a pull request.
```

Author: Done.

## Pull Request Workflow

* If you are in the `contributor` role, you must first fork the repository you wish to make changes in and then submit the pull request to the upstream. If you’re not familiar with pull requests, see the [GitHub pull request documentation](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/about-pull-requests).
* Pull requests should be reviewed by at least one `committer` that is not the PR author before merging.
    * For pull requests created by `committers`, the pull request author is expected to perform the pull request merge after another committer has approved the pull request.
    * For pull requests created by contributors, the committer reviewing the pull request is responsible for merging the pull request after approval.
* Make sure that the latest version of main is branched for the pull request.
* Branch names should include a ticket number and short description.
    * Example: `978-fix-nihms-loader-etl`
* The description in your pull request should include the following:
    * Summary of major changes and what the pull request will accomplish.

>     * Include instructions of how to test the changes.

Reviewer: ```suggestion
    * Instructions identifying how to test the changes.
```

Author: Done.

* Ensure that every PR is linked to a relevant ticket.
* Update or add new documentation to the `pass-documentation` repository.
    * This would be a separate pull request, see the Documentation section for this process.
* A pull request should not be merged unless all automated checks pass.
* Merges should happen using the rebase strategy.

>     * If there have been changes to main, you may want to rebase your branch on main for additional safety.

Reviewer: ```suggestion
    * If there have been changes to the `main` code branch, you may want to rebase your branch on `main` for additional safety.
```

Author: Done.

* After a successful merge, delete the branch.

## Pull Request Review Process

> 

Reviewer: This section seems incomplete without a bullet indicating the need to review the code and what to pay attention to in doing code reviews

Author: I didn't get any additional feedback on this section. I didn't want to add anything that was outside our current team norms. 

I will add some suggestions and send a message to our slack channel to see if anyone has any comments about them.

Author: Done.

* After a successful merge, delete the branch.

## Pull Request Review Process

* When reviewing the code, here are some advised areas to consider:
  * Verify that the implemented logic aligns with the requirements and specifications.
  * Look for any potential bugs or logical errors.
  * Ensure code is adequately commented where necessary.
  * Verify that sensitive configurations are externalized and not hard-coded.
  * Ensure REST endpoints follow standard conventions (e.g., proper use of [HTTP methods](https://developer.mozilla.org/en-US/docs/Web/HTTP)).
* Review any unit/integration tests and ensure they provide proper coverage.
  * Ensure proper use of mocks and stubs to isolate components during testing.
* Identify and suggest refactoring for any [code smells](https://linearb.io/blog/what-is-a-code-smell).
  * Look for areas that could benefit from improved [design patterns or structures](https://www.baeldung.com/design-patterns-series).
* Ensure that dependencies are properly managed and up-to-date.
  * Check for any potential conflicts or unused dependencies.
* Review commit messages.

> * Build the project and run tests locally using `mvn verify` or `mvn clean install`.

Author: That's a good point, it is backend specific. I'm in favor at least this point in time to remove the commands and use a more universal language like `Build the project and run tests locally`.

Personally, for the backend I would always run and build the project locally using `mvn verify`. I've only let acceptance tests run in CI, but perhaps we do want to include them to run locally? Sometimes for certain changes I will do some sanity testing and run pass-docker with the changes. 

Would changing this to `Build the project and run tests locally` be good for now? I'm open to discussing this point further with the entire team too.

> * Build the project and run tests locally using `mvn verify` or `mvn clean install`.

Author: Definitely a good discussion for us as a team. We could discuss this more during the retreat. I will make note of it.

> * Build the project and run tests locally using `mvn verify` or `mvn clean install`.

Author: Done.
